### ISSUE TITLE
Multiple grants override eachother's roles

### ISSUE THREAD

**Issue by ob1jedi** [Snowflake-Labs/terraform-provider-snowflake#1227]
**Provider Version**
Snowflake-Labs/snowflake 0.44.0
 
**Terraform Version**
0.13

**Describe the bug**
Multiple grants on a common resource cause eachother's roles-grants to be deprovisioned.

**Expected behavior**
Expect to provision one grant to one role on one resource, without impacting other roles granted to the same resource.

**Code samples and commands**
```terraform

# create 2 roles for demonstration purposes...
resource snowflake_role RL_TEST_FIRST {
  name = "RL_TEST_FIRST"
}
resource snowflake_role RL_TEST_SECOND {
  name = "RL_TEST_SECOND"
}

# Provision a grant anywhere... (inside tf, outside tf, in the same script, or a different script)
resource snowflake_database_grant db_FIRST {
  depends_on = [snowflake_role.RL_TEST_FIRST]
  database_name     = "TFTEST"
  privilege         = "USAGE"
  roles             = ["RL_TEST_FIRST"]
  with_grant_option = false
  lifecycle {
    ignore_changes = [shares]
  }
}

# Provision the same grant for another role anywhere other than inside the first resource... (inside tf, outside tf, in the same script, or a different script)
resource snowflake_database_grant db_SECOND {
  depends_on = [snowflake_role.RL_TEST_SECOND]
  database_name     = "TFTEST"
  privilege         = "USAGE"
  roles             = ["RL_TEST_SECOND"]
  with_grant_option = false
  lifecycle {
    ignore_changes = [shares]
  }
}
```

Running more than once produces the following constantly...
```terraform
 # snowflake_database_grant.db_FIRST must be replaced
-/+ resource "snowflake_database_grant" "db_FIRST" {
      ~ id                     = "TFTEST|||USAGE|RL_TEST_FIRST|false" -> (known after apply)
      ~ roles                  = [ # forces replacement
          - "RL_TEST_SECOND",
            # (1 unchanged element hidden)
        ]
      - shares                 = [] -> null
        # (4 unchanged attributes hidden)
    }

  # snowflake_database_grant.db_SECOND must be replaced
-/+ resource "snowflake_database_grant" "db_SECOND" {
      ~ id                     = "TFTEST|||USAGE|RL_TEST_SECOND|false" -> (known after apply)
      ~ roles                  = [ # forces replacement
          - "RL_TEST_FIRST",
            # (1 unchanged element hidden)
        ]
      - shares                 = [] -> null
        # (4 unchanged attributes hidden)
    }
```

**Additional context**
I've tried bundling in all the existing roles, but then removing one resource with an associated role ends up destroying all the role grants.
I've tried adding "ignore roles" but when I deprovision any grant resource, it still destroys the grants for all other roles, regardless of where they were created.
I've refactored my code so many times now to try to deal with this issue that I'm just about over it. Please help.

**Comment by frsann** [comment]
Try the `enable_multiple_grants` option: [Docs](https://registry.terraform.io/providers/Snowflake-Labs/snowflake/latest/docs/resources/database_grant#enable_multiple_grants)

**Comment by miditha** [comment]
@frsann I tried with `enable_multiple_grants` too, but still it revokes the grants which was applied previously.

In my case
```
# First grant USAGE to a role
resource "snowflake_database_grant" "grant_role" {
  database_name = "DB1"

  roles     = ["ROLE_1"]
  privilege = "USAGE"

  enable_multiple_grants = true

}

# Then grant USAGE to a share
resource "snowflake_database_grant" "grant_share" {
  database_name = "DB1"

  shares    = ["SHARE_1"]
  privilege = "USAGE"

  enable_multiple_grants = true
}
```
When I run share grant, it revokes the role grant and apply the share grant only.
I noticed the same behaviour with schema and table grants too

**Comment by frsann** [comment]
We ran into similar issue, and the issue for us was that the `enable_multiple_grants` flag was unset in the state, causing the plans to still include the revocation. What we ended up doing was going into the state and adding the flag manually. Yes, very ugly, but did the trick. And since then we always set the `enable_multiple_grants` flag on new resources.

**Comment by miditha** [comment]
Thanks @frsann. I checked the state and I can see `enable_multiple_grants` is set to true

```
    {
      ...
      "mode": "managed",
      "type": "snowflake_database_grant",
      "name": "grant_share",
      "provider": "provider[\"registry.terraform.io/snowflake-labs/snowflake\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "database_name": "DB1",
            "enable_multiple_grants": true,
            "id": "DB1|||USAGE||false",
            "privilege": "USAGE",
            "roles": null,
            "shares": [
              "SHARE1"
            ],
            "with_grant_option": false
          },
          "sensitive_attributes": [],
          ...
          "dependencies": [
          ...
          ]
        }
      ]
    }
```
Still it tries to revoke the role privileges which were granted previously

**Comment by frsann** [comment]
Alright, then I’m out of ideas, sorry. 😔

**Comment by systematics78** [comment]
I have facing the same issue. My task is grant usage access to share.. so it revokes the role grants. enable_multiple_grants set true from beginning. provider version 0.47.0 Any idea how to resolve?

**Comment by matt-macnish-cko** [comment]
This is also happening with some of our grants.  However this is only happening in some environments.  We are using a common module for our deployments so the only difference between environments are some names.

**Comment by matt-macnish-cko** [comment]
I believe this is caused by when the API reaches out it then populates state with all the roles that the priviledge is assigned.  Then when a later run happens it checks this against the code and sees none of those roles and marks them for deletion.  I have confirmed that state is populated with Roles not in the code.  Enable_multiple_grants is set to true as well in state.

**Comment by fernandobrito** [comment]
@systematics78 I believe your problem is caused because `enable_multiple_grants` seems to have no effect for `shares`. Since this GitHub issue here has now different people reporting different things, I just created a new one more specific and descriptive: https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/1339.

**Comment by fernandobrito** [comment]
Note that there's another issue: if you use the grant resource to give privileges to shares and leave the `roles` arguments empty, it will remove all the roles privileges, even if `enable_multiple_grants` is true. I also created an issue for it: https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/1341.

**Comment by sfc-gh-swinkler** [comment]
fixed in pr #1510 . Will be released as part of 0.56.3
